### PR TITLE
pset: add optional asset blinding factor to input and output

### PR DESCRIPTION
**Pull request by LeoComandini**
In some protocols where blinding is done by different parties,
the asset surjection proof for some output might be created by:
* someone who does not own an input with the same asset, or
* someone who does not know the asset blinding factor
  corresponding to the output asset commitment

This, in general, is the case of LiquiDEX v1 [1], e.g.
* Alice wants to swap `x` of asset `A` for `y` of asset `B`
* Alice owns a utxo with `x` of asset `A`
* Alice creates a spending that utxo and receiving an output
  with `y` of asset `B`.
* Alice draws at random the asset and value blinding factors
  for the output, sets the corresponding asset and value
  commitment. Sets the nonce commitment. Computes and sets
  the rangeproof.
* Alice computes the scalar offset contribution of the input
  and output (combined)
* Alice computes the value blind proofs for the input and output
* Alice creates the LiquiDEX proposal using the data obtained
  above:
```
{
  "version": 1,
  "tx": "...",
  "inputs": [{
    "asset": "aa...",
    "satoshi": x,
    "assetblinder": "...",
    "value_blind_proof": "...",
  }],
  "outputs": [{
    "asset": "bb...",
    "satoshi": y,
    "assetblinder": "...",
    "value_blind_proof": "...",
  }],
  "scalars": ["..."],
}
```
* Alice shares the proposal with Bob
* Bob adds more inputs for the asset `B` and fees
* Bob adds more outputs for the asset `A`, `B` and fees 
* Bob blinds the transaction, i.e.:
  * draws at random abf and vbf for each new output, apart 
    from the last one for which he uses the new inputs 
    contribution and the scalar offset from the proposal to 
    balance the tx.
  * creates rangeproofs for each new output
  * creates surjection proofs for each (blinded) fee output 
  * creates surjection proofs for each `A` output, note that 
    in general the input asset blinding factor is needed.
  * creates surjection proofs for each `B` output, including 
    the one from Alice, which requires the output blinding 
    factor. Note that Alice could not have created the
    surjection proof since she did not know any `B` input
    when she created the tx.

Input and output asset blinding factors are the last fields 
needed to convert LiquiDEX v1 proposals in PSETs (and viceversa).

---

[1] https://leocomandini.github.io/2022/10/27/liquidexv1.html

### CODE REVIEW

File: doc/pset.mediawiki [hunk 1]
```diff
@@ -305,6 +305,16 @@ The currently defined elements per-input proprietary types are as follows:
 |
 | 0
 | 2
+|-
+| Asset Blinding Factor
+| <tt>PSBT_ELEMENTS_IN_ASSET_BLINDING_FACTOR = 0x16</tt>
+| None
+| No key data
+| <tt><proof></tt>
```
Comment (delta1) [inline doc/pset.mediawiki:313]: leftover from copy/paste? 

```suggestion
| <tt><256 bit uint></tt>
```

File: doc/pset.mediawiki [hunk 2]
```diff
@@ -418,6 +429,16 @@ The currently defined elements per-output proprietary types are as follows:
 |
 | 0
 | 2
+|-
+| Asset Blinding Factor
+| <tt>PSBT_ELEMENTS_OUT_ASSET_BLINDING_FACTOR = 0x0b</tt>
+| None
+| No key data
+| <tt><proof></tt>
```
Comment (delta1) [inline doc/pset.mediawiki:437]: same here

File: src/psbt.h [hunk 3]
```diff
@@ -1096,6 +1105,18 @@ struct PSBTInput
                                 m_blinded_issuance = b;
                                 break;
                             }
+                            case PSBT_ELEMENTS_IN_ASSET_BLINDING_FACTOR:
+                            {
+                                if (m_asset_blinding_factor.has_value()) {
+                                    throw std::ios_base::failure("Duplicate Key, asset blinding factor is already provided");
```
Comment (delta1) [inline src/psbt.h:1111]:
```suggestion
                                    throw std::ios_base::failure("Duplicate Key, input asset blinding factor is already provided");
```

**Review by delta1 — APPROVED** [review]
utACK bc3870a06fae4763afdc0f4baa3493065619d231 with minor comments

File: doc/pset.mediawiki (hunk 1 above)
Comment (LeoComandini) [reply doc/pset.mediawiki]: changed

File: doc/pset.mediawiki (hunk 2 above)
Comment (LeoComandini) [reply doc/pset.mediawiki]: changed

File: src/psbt.h (hunk 3 above)
Comment (LeoComandini) [reply src/psbt.h]: added, thanks

**Comment by LeoComandini** [comment]
Hi @delta1 ,
I've addressed your comments and rebased.
Should I do something more to have this merged?

**Comment by delta1** [comment]
hey @LeoComandini, I was hoping to get another set of eyes on this. 

@apoelstra could you please review?

**Comment by apoelstra** [comment]
Does anyone remember where the (very long past discussion was where I was opposed to putting raw secret data into PSETs?

I think this should have an ELIP number.

**Comment by LeoComandini** [comment]
> Does anyone remember where the (very long past discussion was where I was opposed to putting raw secret data into PSETs)?

Pinged you in that chat

You also reviewed and merged the same change in rust-elements, see https://github.com/ElementsProject/rust-elements/pull/201

**Comment by apoelstra** [comment]
> You also reviewed and merged the same change in rust-elements, see [ElementsProject/rust-elements#201](https://github.com/ElementsProject/rust-elements/pull/201)

I had **assumed** that PRs to rust-elements from Blockstream employees which add PSET fields with no description or justification were implementing missing functionality. Not that they were controversial out-of-spec extensions. I'll review more carefully in the future.

**Comment by jsarenik** [comment]
> Does anyone remember where the (very long past discussion was where I was opposed to putting raw secret data into PSETs?
> 
> I think this should have an ELIP number.

@apoelstra please share as I could not find it yet.

Some quick candidates may be:

 * https://github.com/bitcoin-core/secp256k1/pull/262
 * https://github.com/BlockstreamResearch/secp256k1-zkp/pull/46
 * https://github.com/rust-bitcoin/rust-bitcoin/pull/2770
 * https://github.com/rust-bitcoin/rust-miniscript/issues/483

**Comment by LeoComandini** [comment]
@apoelstra , opened ELIP MR https://github.com/ElementsProject/ELIPs/pull/18

**Comment by apoelstra** [comment]
Merged the ELIP.

Since this is LiquiDEX-specific and a proprietary extension, I think we should close this PR and instead only implement it in rust-elements/elements-miniscript.

**Comment by delta1** [comment]
Closing since this is now in rust-elements https://github.com/ElementsProject/rust-elements/pull/207  

@LeoComandini if you do need this in Elements then let's revisit this discussion

**Pull request closed without merging**